sensor/common/registry/tlscheck_cache.go

```diff
@@ -0,0 +1,98 @@
+package registry
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/pkg/env"
+	"github.com/stackrox/rox/pkg/expiringcache"
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+type tlsCheckResult uint8
+
+const (
+	tlsCheckResultUnknown tlsCheckResult = iota
+	tlsCheckResultSecure
+	tlsCheckResultInsecure
+	tlsCheckResultError
+)
+
+var (
+	tlsCheckTTL = env.RegistryTLSCheckTTL.DurationSetting()
+)
+
+type tlsCheckCacheImpl struct {
+	// results holds results from TLS checks. This prevents repeatedly
+	// performing checks on the same registry within the cache expiry
+	// window. An expiring cache is used because the TLS state of a registry
+	// may change.
+	results      expiringcache.Cache
+	checkTLSFunc CheckTLS
+}
+
+type cacheEntry struct {
+	once   sync.Once
+	result tlsCheckResult
+}
+
+func (e *cacheEntry) checkTLS(ctx context.Context, registry string, checkTLSFunc CheckTLS) (secure bool, skip bool, err error) {
+	e.once.Do(func() {
+		secure, err = checkTLSFunc(ctx, registry)
+		if err != nil {
+			e.result = tlsCheckResultError
+			return
+		}
+
+		e.result = tlsCheckResultInsecure
+		if secure {
+			e.result = tlsCheckResultSecure
+		}
+	})
+
+	if err != nil {
+		return false, false, err
+	}
+
+	switch e.result {
+	case tlsCheckResultSecure:
+		return true, false, nil
+	case tlsCheckResultInsecure:
+		return false, false, nil
+	case tlsCheckResultError:
+		return false, true, nil
```

Comment:
why skip here when running into the error the first time does not skip? (`true` vs `false` for `skip`)

Reply:
I think it's explained below:

```
[...] Returns true for skip if there was a previous error and the
// registry should not be upserted into the store.
```

So we do upsert the error, but only the first time?

Can we add a comment to this function, too, because when reviewing this file I didn't see the `CheckTLS` until after I saw this